keps/sig-apps/20190226-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,231 @@
+further complicate the situation(I know they were solved). StatefulSet introduced ControllerRevisions in 1.7 which I believe are easier to think and reason 
+about. They are used by DaemonSet and StatefulSet for tracking revisions. It would be so much nicer if all the use cases of Deployments can be met and we 
+could track the revisions by ControllerRevisions.
+
+With this feature in place, when using StatefulSet with maxUnavailable >1, the user understands that this would not cause issues with their Stateful 
+Applications which have per pod state and identity while still providing all of the above written advantages.
+
+### Goals
+StatefulSet RollingUpdate strategy will contain an additional parameter called `maxUnavailable` to control how many Pods will be brought down at a time,
+during Rolling Update.
+
+### Non-Goals
+maxUnavailable is only implemeted to affect the Rolling Update of StatefulSet. Considering maxUnavailable for Pod Management Policy of Parallel is beyond 
+the purview of this KEP.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a User of Kubernetes, I should be able to update my StatefulSet, more than one Pod at a time, in a RollingUpdate way, if my Stateful app can tolerate 
+more than one pod being down, thus allowing my update to finish much faster. 
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the Rolling Update Strategy for StatefulSet.
+
+```go
+// RollingUpdateStatefulSetStrategy is used to communicate parameter for RollingUpdateStatefulSetStrategyType.
+type RollingUpdateStatefulSetStrategy struct {
+	// THIS IS AN EXISTING FIELD
+        // Partition indicates the ordinal at which the StatefulSet should be
+        // partitioned.
+        // Default value is 0.
+        // +optional
+        Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
+
+	// NOTE THIS IS THE NEW FIELD BEING PROPOSED
+	// The maximum number of pods that can be unavailable during the update.
+        // Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+        // Absolute number is calculated from percentage by rounding down.
+        // Defaults to 1.
+        // +optional
+        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
+	
+	...
+}
+```
+
+- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets will follow their old behavior. This
+  will also help while upgrading from a release which doesnt support maxUnavailable to a release which supports this field.
+- If maxUnavailable is specified, it cannot be greater than total number of replicas.
+- If maxUnavailable is specified and partition is also specified, MaxUnavailable cannot be greater than `replicas-partition`
+- If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the partition. Which means all Pods 
```

Comment:
@kow3ns @janetkuo @kargakis  @FillZpp i have added three options below for more discussions, please review.

Reply:
pinging again @kow3ns  @janetkuo  who should i follow up to get this merged and should the followup discussion happen in a new  PR or a issue for further design details ?